plugins/azure-devops/src/utils/getAnnotationValuesFromEntity.ts

```diff
@@ -0,0 +1,125 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  AZURE_DEVOPS_BUILD_DEFINITION_ANNOTATION,
+  AZURE_DEVOPS_HOST_ORG_ANNOTATION,
+  AZURE_DEVOPS_PROJECT_ANNOTATION,
+  AZURE_DEVOPS_REPO_ANNOTATION,
+} from '../constants';
+
+import { Entity } from '@backstage/catalog-model';
+
+export function getAnnotationValuesFromEntity(entity: Entity): {
+  project: string;
+  repo?: string;
+  definition?: string;
+  host?: string;
+  org?: string;
+} {
+  const { host, org } = getHostOrg(entity.metadata.annotations);
+
+  const projectRepoValues = getProjectRepo(entity.metadata.annotations);
+  if (projectRepoValues.project && projectRepoValues.repo) {
+    return {
+      project: projectRepoValues.project,
+      repo: projectRepoValues.repo,
+      host,
+      org,
+    };
+  }
+
+  const project =
+    entity.metadata.annotations?.[AZURE_DEVOPS_PROJECT_ANNOTATION];
+  if (!project) {
+    throw new Error(
+      `Value for annotation ${AZURE_DEVOPS_PROJECT_ANNOTATION} was not found`,
+    );
+  }
+
+  const definition =
+    entity.metadata.annotations?.[AZURE_DEVOPS_BUILD_DEFINITION_ANNOTATION];
+  if (!definition) {
+    throw new Error(
+      `Value for annotation ${AZURE_DEVOPS_BUILD_DEFINITION_ANNOTATION} was not found`,
+    );
+  }
+  return { project, definition, host, org };
+}
+
+function getProjectRepo(annotations?: Record<string, string>): {
+  project?: string;
+  repo?: string;
+} {
+  const annotation = annotations?.[AZURE_DEVOPS_REPO_ANNOTATION];
+  if (!annotation) {
+    return { project: undefined, repo: undefined };
+  }
+
+  if (!annotation.includes('/')) {
+    throw new Error(
+      `Value for annotation ${AZURE_DEVOPS_REPO_ANNOTATION} was not in the correct format: <project-name>/<repo-name>`,
+    );
+  }
+
+  const [project, repo] = annotation.split('/');
+
+  if (!project) {
+    throw new Error(
+      `Project Name for annotation ${AZURE_DEVOPS_REPO_ANNOTATION} was not found; expected format is: <project-name>/<repo-name>`,
+    );
+  }
+
+  if (!repo) {
+    throw new Error(
+      `Repo Name for annotation ${AZURE_DEVOPS_REPO_ANNOTATION} was not found; expected format is: <project-name>/<repo-name>`,
+    );
+  }
+
+  return { project, repo };
+}
+
+function getHostOrg(annotations?: Record<string, string>): {
+  host?: string;
+  org?: string;
+} {
+  const annotation = annotations?.[AZURE_DEVOPS_HOST_ORG_ANNOTATION];
+  if (!annotation) {
+    return { host: undefined, org: undefined };
+  }
+
+  if (!annotation.includes('/')) {
+    throw new Error(
+      `Value for annotation ${AZURE_DEVOPS_HOST_ORG_ANNOTATION} was not in the correct format: <host-name>/<organization-name>`,
```

Comment:
Suggestion: This error message could do well to include the actual value found; that makes the source of the error easier to locate. Could also refactor to have positive tests and a single throw

```ts
if (annotation.includes('/')) {
  if (...) {
    return;
  }
}

throw new Error(`Invalid value for annotation '${AZURE_DEVOPS_HOST_ORG_ANNOTATION}'; expected format X, found '${annotation}'`);
```

Same for the others above

Reply:
Good call, cleaned up the code as well, thanks!